src/Mono.Android/metadata

```diff
@@ -1884,6 +1884,43 @@
   <!-- These methods are hand-bound in HardwareBuffer.cs because they have a long enum type, which generator doesn't support. -->
   <remove-node api-since="26" path="/api/package[@name='android.hardware']/class[@name='HardwareBuffer']/method[@name='create' and count(parameter)=5 and parameter[1][@type='int'] and parameter[2][@type='int'] and parameter[3][@type='int'] and parameter[4][@type='int'] and parameter[5][@type='long']]" />
   <remove-node api-since="26" path="/api/package[@name='android.hardware']/class[@name='HardwareBuffer']/method[@name='getUsage' and count(parameter)=0]" />
+	
+  <!-- These are new abstract methods in existing classes, so they must be converted to virtual compatible method for back-compat. -->
+  <attr api-since="35" path="/api/package[@name='java.nio']/class[@name='MappedByteBuffer']/method[@name='compact' and count(parameter)=0]" name="managedOverride">override</attr>
+  <attr api-since="35" path="/api/package[@name='java.nio']/class[@name='MappedByteBuffer']/method[@name='compact' and count(parameter)=0]" name="compatVirtualMethod">true</attr>
+
+  <!-- Rename Namespaces -->
+  <attr api-since="35" path="/api/package[@name='android.adservices.ondevicepersonalization']" name="managedName">Android.AdServices.OnDevicePersonalization</attr>
+  <attr api-since="35" path="/api/package[@name='android.adservices.signals']" name="managedName">Android.AdServices.Signals</attr>
+  <attr api-since="35" path="/api/package[@name='java.util.random']" name="managedName">Java.Util.RandomGenerators</attr>
+
+  <!-- API-35 adds new java.util.Sequenced[Collection|Map] interfaces that existing interfaces/classes now implement. -->
+  <!-- Fix covariant return types for new 'reversed ()' method. -->	
+  <attr api-since="35" path="/api/package[@name='java.util']/interface[@name='List']/method[@name='reversed' and count(parameter)=0]" name="managedReturn">Java.Util.ISequencedCollection</attr>
```

Comment:
This feels like we'll wind up having *two* `Reversed` methods in the inheritance tree, a'la:

```csharp
interface ISequencedCollection {
    ISequencedCollection Reversed();
}
interface List : ISequencedCollection {
    ISequencedCollection Reversed();
}
```

If correct, shouldn't we instead "reabstract" `IList.Reversed()`? https://github.com/xamarin/Java.Interop/commit/22d5687bc5c572ea0eadbd1a23f02f424af782d1

…and ditto most of the subsequent `managedReturn` changes?

Reply:
This is how it looks in Java:

```java
public interface SequencedCollection<E> extends Collection<E> {
  SequencedCollection<E> reversed();
}

public interface List<E> extends SequencedCollection<E>, Collection<E> {
  default List<E> reversed() {
    throw new RuntimeException("Stub!");
  }
}
```

This is just a normal "provide a `default` implementation for an inherited interface method".

I think reabstraction is for the opposite case: when we want to "hide" a `default` implementation and make it `abstract` again.

Reply:
Aside: This is a matter of "leaky abstraction I'd rather not make worse".  C# permits "explicitly implementing interfaces", which is *required* for things like `IEnumerable.GetEnumerator()` vs. `Enumerable<T>.GetEnumeartor()` (same name! Different semantics! Supported via explicitly implementing interface members).

Java does not have an equivalent to explicit interface implementation.  All method names must have the same semantics and "compatible" return types.

This is a "leaky abstraction" that we've long allowed to happen, largely because we had no way to address it other than `<remove-node/>` (which perhaps we should have done?), which results in things like:

```csharp
namespace Java.Util;
partial interface ICollection {
    bool IsEmpty {get;}
    bool Add (Object? e);
    // …
}
partial interface IList : ICollection {
    // same methods
    bool IsEmpty {get;}
    bool Add (Object? e);
}
```

Things are "fine" so long as you *don't* explicitly implement them.  If you *do*:

```csharp
class MyList : Java.Lang.Object, Java.Util.IList {
    bool Java.Util.ICollection.IsEmpty { get => true; }
    bool Java.Util.IList.IsEmpty { get => false; }
}
```

which do you get?  There Can Be Only One™.

Answer: `IList.IsEmpty`, because it's the most derived type, as per the JCW:

```java
		__md_methods = 
			"n_isEmpty:()Z:GetIsEmptyHandler:Java.Util.IListInvoker, Mono.Android, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null\n" +
```

So it "works", but is not necessarily ideal.

Thus, the question: *What **is** ideal*, and can we get closer to it?

Given that our current C# approach results in numerous C# warnings that we have to *ignore* ([CS0114](https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs0114)), I think that the ideal scenerio would be to "reabstract" the members:

```csharp
namespace Java.Util;
interface IList : ICollection {
    abstract bool ICollection.Add(Java.Lang.Object? e);
    abstract bool ICollection.IsEmpty {get;}
}
```

This would allow you to explicitly implement e.g. *only* `ICollection.Add()`, and *not* `IList.Add()`, as they're now "flattened"/"unified".

We can't do this for existing members (API & ABI break).

We *can* do this for *newly added* members, if/when appropriate.

On to `default` method implementations: *C#* supports this as well, again via explicit interface implementation:

```csharp
namespace Java.Util;
interface ICollection {
    // has a body for back-combat reasons
    ISequencedCollection Reversed() => throw new NotImplementedException();
}
interface IList : ICollection {
    IList ICollection.Reversed() => _members.InstanceMethods.InvokeAbstracMethod(…);
}
```

Reply:
On Discord, @jpobst [asks](https://discord.com/channels/732297728826277939/732297837953679412/1229904825307758612):

> i'm not sure if your latest comment is suggesting a change? or just opining? 😁

Er, yes.

Background info, plus opinion on where we should consider going, neither of which is directly relevant for `.Reversed()`.  The comment is thus not *actionable*.

My bad.

***However***, now that I have that background thinking in place, it allows me to point to:

  * [`SequencedCollection`](https://developer.android.com/reference/java/util/SequencedCollection) and/vs
  * [`List`](https://developer.android.com/reference/java/util/List)

*Both* declare `getFirst()`/`getLast()`/`removeFirst()`/`removeLast()`/`reversed()` methods, *suggesting* -- as I haven't build this PR locally, I just believe that this will likely be the case -- that the binding with this PR as-is will contain:

```csharp
namespace Java.Util;
partial interface ISequencedCollection {
    void AddFirst(Object? e);
    void AddLast(Object? e);
    Object? GetFirst();
    Object? GetLast();
    // …
}

partial interface IList : ISequencedCollection {
    void AddFirst(Object? e);  // CS0114 warning
    void AddLast(Object? e);   // CS0114 warning
    Object? GetFirst();        // CS0114 warning
    Object? GetLast();         // CS0114 warning
    // …
}
```

and I ask: *should it be this way*?

It's *been* this way, so it's not "broken" per-se, but it doesn't feel "entirely proper" either.

There are two solutions for making this "more proper":

 1. Use `<remove-node/>` to remove the of identically declared methods from `List`.
 2. Use `reabstract` metadata to declare the `IList` methods as `abstract` and explicltly declare them.

Or (3), update `generator` to internally handle (1) or (2). :-). 
Which becomes all the more important/required given that *we don't know* how many other interfaces this impacts (though `SequencedMap`/`Map` & `SequencedSet`/`Set` also hit it, minimum).  (For comparison, if I stop ignoring CS0114 in `Java.Base`, there are *61 warnings* emitted that I now feel *shouldn't* be emitted, and that's a *very small* subset of `java.base.jmod`…)

This comment is thus deeply in "the perfect is the enemy of the good" territory, mind you…

Reply:
…so where does that leave me?

If we were doing it all over again -- *cough* `Java.Base` *cough* -- then I'd want the CS0114 warnings to *not* be emitted by tooling, *either* by removing identically declared overriding members *or* by re-abstracting them.

However, we're *not* doing it all over again, we have no tooling to assist in finding these situations (other than re-enabling CS0114, which would introduce *686* warnings!), so manually fixing these is a non-starter.  We could update `generator`, but I'm not certain that we have the time, *and* this would need to be an "opt-in" construct, somehow, so that we don't re-abstract or remove previously declared methods (though API-compat is a good backstop for this).

I thus think that we can't/shouldn't attempt to address this right now. :-(

Reply:
Attempted to clarify my thinking here: https://github.com/xamarin/java.interop/issues/1215

Reply:
I think my sticking point with understanding this is that the new methods defined in `IList` are `default` methods, so I'm not sure that I comprehend marking them `abstract` with reabstraction.

IE:

```csharp
namespace Java.Util;
partial interface ISequencedCollection {
    void AddFirst(Object? e);
    void AddLast(Object? e);
    Object? GetFirst();
    Object? GetLast();
    // …
}

partial interface IList : ISequencedCollection {
    void AddFirst(Object? e) { // implementation };  // CS0108 warning
    void AddLast(Object? e); { // implementation };   // CS0108 warning
    Object? GetFirst(); { // implementation };       // CS0108 warning
    Object? GetLast(); { // implementation };         // CS0108 warning
    // …
}
```

Aside, it's a `CS0108` warning instead of `CS0114`.

```
CS0108 'IList.AddFirst(Object?)' hides inherited member 'ISequencedCollection.AddFirst(Object?)'. Use the new keyword if hiding was intended.
```

I think `CS0108` is the point you are getting at: we aren't "providing an default implementation of `ISequencedCollection.AddFirst`", we are declaring a new `AddFirst` that hides the original one.

However we appear to actually handle this correctly for the specific case where a DIM is providing an implementation for an inherited interface method, as this is what we generate:

```csharp
partial interface IList : ISequencedCollection {
    void Java.Util.AddFirst(Object? e) { // implementation };
    void Java.Util.AddLast(Object? e); { // implementation };
    Object? Java.Util.GetFirst(); { // implementation };
    Object? Java.Util.GetLast(); { // implementation };
    // …
}
```

This prevents the `CS0108` warning and provides the correct "provide a default implementation" semantics rather than hiding the inherited method.

Looking at _other_ existing methods in `IList`/`ICollection` I do see what you are talking about:

```csharp
public partial interface ICollection
    bool AddAll (System.Collections.ICollection c);
}

public partial interface IList : Java.Util.ICollection {
    bool AddAll (System.Collections.ICollection c);  // CS0108
}
```
Fixing that (which we do not do today) indeed would be:

```csharp
abstract bool ICollection.AddAll (System.Collections.ICollection c);
```

To your point, without a way to detect these cases currently, we do not know if API-35 adds any additional cases that we could manually fix with reabstraction or removal.